### DISCUSSION TITLE
Content Security Policy - Refused to frame

### DISCUSSION THREAD

**Discussion by PedroReyes** [WalletConnect/walletconnect-monorepo#4341] [Developer Support] [answered]
Hello @glitch-txs  ,

I am creating here the ticket as you suggested in discord (my issue is coming back to [here](https://discord.com/channels/492410046307631105/1196875414572900363/1212573552130396191). I transcript it here:

I have verified my app in WalletConnect but still when I click the button, the modal does not open in production environment nor it does on localhost:3000. 

I have updated my dependencies as you suggested [here](https://discord.com/channels/492410046307631105/1196875414572900363/1212592196113539132):

```json 
// 📖 package.json
  "dependencies": {
    "@web3modal/ethers": "3.5.7",
    "ethers": "6.6.3",
    "next": "14.0.1",
    "react": "18.2.0",
    "react-dom": "18.2.0",
  }
```

```js
// 📖 next.config.js

/** @type {import('next').NextConfig} */
const nextConfig = {
  webpack: (config) => {
    config.externals.push("pino-pretty", "lokijs", "encoding");
    return config;
  },
  minifySWC: false,
};

module.exports = nextConfig;
```

After these updates I am still getting the next error:

```js
Refused to frame 'https://verify.walletconnect.org/' because an ancestor violates the following Content Security Policy directive: "frame-ancestors https://*.myapp.web.app https://myapp.web.app"
```

Full console error here:

```js
Refused to frame 'https://verify.walletconnect.org/' because an ancestor violates the following Content Security Policy directive: "frame-ancestors https://*.myapp.web.app https://myapp.web.app"

swap:1  Third-party cookie will be blocked. Learn more in the Issues tab.
swap:1  Third-party cookie will be blocked. Learn more in the Issues tab.
swap:1  Third-party cookie will be blocked. Learn more in the Issues tab.

chromewebdata/:2157  Uncaught TypeError: Cannot read properties of null (reading 'classList')
    at onResize (chromewebdata/:2157:38)
    at HTMLDocument.setupMobileNav (chromewebdata/:2184:3)
onResize @ (index):2157
setupMobileNav @ (index):2184

chromewebdata/:2745  Uncaught TypeError: Cannot read properties of null (reading 'classList')
    at HTMLDocument.onDocumentLoad (chromewebdata/:2745:19)
onDocumentLoad @ (index):2745

Refused to frame 'https://verify.walletconnect.org/' because an ancestor violates the following Content Security Policy directive: "frame-ancestors https://*.myapp.web.app https://myapp.web.app".

(index):2157  Uncaught TypeError: Cannot read properties of null (reading 'classList')
    at onResize ((index):2157:38)
    at HTMLDocument.setupMobileNav ((index):2184:3)

onResize @ (index):2157
setupMobileNav @ (index):2184
(index):2745  Uncaught TypeError: Cannot read properties of null (reading 'classList')
    at HTMLDocument.onDocumentLoad ((index):2745:19)
```
I have changed the real name of my app to "myapp" in the examples below.

I have added all the URLs (and more) that appears in the error to my allowed domains:

![image](https://github.com/WalletConnect/walletconnect-monorepo/assets/4096108/4e7d4dc1-673c-430a-9aa7-a3d1d59287a4)

I don't come up with any other thing that might be causing my error, thereby my ask for help 🙂

Thanks in advanced for your help. Please, do not hesiste to contact me on disccord: `pedroreyessan`

Best regards,
Pedro Reyes.

PD: Regarding the update in the `next.config.js` I am starting to get this warning in my logs:

```js
 ⚠ Invalid next.config.js options detected:
 ⚠     Unrecognized key(s) in object: 'minifySWC'
 ⚠ See more info here: https://nextjs.org/docs/messages/invalid-next-config
```

**Comment by glitch-txs** [comment]
hi could you update to v4? also could you share your metadata configuration in the code? Thanks!

**Reply by PedroReyes** [reply]
Hello @glitch-txs ,

Thanks for the prompt reply!

I have updated to `"@web3modal/ethers": "4.1.1",`

And here it is my metadata configuration:

```typescript
"use client";

import { createWeb3Modal, defaultConfig } from "@web3modal/ethers/react";

// 1. Get projectId at https://cloud.walletconnect.com
const projectId = "****************"; // my dev/test wallet

// 2. Set chains
const bscTestnet = {
  chainId: 97,
  name: "Binance Smart Chain Testnet",
  currency: "BNB",
  explorerUrl: "https://testnet.bscscan.com",
  rpcUrl: "https://data-seed-prebsc-1-s1.binance.org:8545",
};

const bscMainnet = {
  chainId: 56,
  name: "Binance Smart Chain",
  currency: "BNB",
  explorerUrl: "https://bscscan.com",
  rpcUrl: "https://bsc-dataseed.binance.org",
};

// 3. Create modal
const domain =
  process.env.NODE_ENV && process.env.NODE_ENV === "development"
    ? "localhost:3000"
    : "https://my-app.web.app";

const metadata = {
  name: "MyApp",
  description: "whatever description you want to add here",
  url: domain,
  icons: [domain + "/static/favicon.ico"],
};

createWeb3Modal({
  ethersConfig: defaultConfig({ metadata }),
  chains: [
    // mainnet,
    bscTestnet,
    bscMainnet,
  ],
  projectId,
  enableAnalytics: true, // Optional - defaults to your Cloud configuration
});

export function Web3ModalProvider({ children }: { children: React.ReactNode }) {
  return children;
}

```

After this updates the problem persists.

**Reply by glitch-txs** [reply]
could you check you're following these recommendations: https://docs.walletconnect.com/cloud/relay#allowlist

**Reply by PedroReyes** [reply]
Yes, I am following. My app is just (example) `my-app.web.app` and I have in allowed domains: `my-app.web.app`

Right now this is the status:

❌ localhost PC: 

display the pop up and connect ❌ the error in the console persist

❌ localhost mobile: 

popping up `WalletConnect` and `Metamask` do not pop up, thereby no connection is done. 

❌production PC: 

not popping up

❌ production mobile: 

not popping up

**Reply by glitch-txs** [reply]
Are you able to share your domain? I can take a look

**Reply by PedroReyes** [reply]
https://firelabs-app.web.app/

**Reply by glitch-txs** [reply]
I'm not seeing the errors on console, how are you opening the modal? do you have any hydration errors locally?

**Reply by PedroReyes** [reply]
Hello @glitch-txs ,

When you click the button what I call for opening the modal is the next code:

```typescript
await get().web3ModalOpenWallet({ view: "Connect" });
```

This function is nothing else but the hook from WalletConnect:

```typescript
const { open } = useWeb3Modal();

. . . 

React.useEffect(() => {
    . . .
    // Set web3modal to open web3 wallet modal
    setWeb3ModalConnectWallet(open);

    . . .
},[]);
```

The rest of the layout component where this `setWeb3ModalConnectWallet` is established is the next:

```html
export default function Layout(){
  // . . . 
  return (
    <React.Fragment>
      <html lang={USER_LANGUAGE} className="bg-[#0A0A0A] h-screen">
        <Head>
          <link rel="icon" sizes="76x76" href="static/favicon.svg" />
        </Head>

        <body className={"h-dvh"} ref={newRef}>
          <Web3ModalProvider>
            <div>
            <Header />
            <main>
              <ToastContainer
                toastStyle={{ backgroundColor: "#0E0E11", color: "#DCDCDD" }}
              />
            </main>
          </Web3ModalProvider>
        </body>
      </html>
    </React.Fragment>
  );
}
```

Here I attach a full video of the status in localhost:

https://github.com/WalletConnect/walletconnect-monorepo/assets/4096108/56b5e390-65d2-4582-a606-c788fe4c5907

Thanks in advanced for helping me out @glitch-txs

**Reply by PedroReyes** [reply]
Hello @glitch-txs ,

what do you think?

**Reply by glitch-txs** [reply]
Hi, I'm not sure how it was structured there, but I'd suggest following one of our examples: https://github.com/WalletConnect/web-examples/tree/main/dapps/web3modal

**Reply by PedroReyes** [reply]
This example in Next.js is for Wagmi. The only difference I see that do not exist in ethers is this piece of code:

```js
<WagmiProvider config={config} initialState={initialState}>
   <QueryClientProvider client={queryClient}>{children}</QueryClientProvider>
</WagmiProvider>
```

Besides that I am following the example @glitch-txs 

Do you have one that uses ethers instead of wagmi?

**Reply by PedroReyes** [reply]
This is the documentation that WalletConnect has for Next.js and ethers (the one I have followed): https://docs.walletconnect.com/web3modal/nextjs/about

Is there something wrong with it that should be updated?

**Reply by glitch-txs** [reply]
could you set up a minimal reproducible example?

**Reply by PedroReyes** [reply]
Yes, I am on it. Give me a few minutes ✌️

**Reply by PedroReyes** [reply]
Here it is @glitch-txs , same errors with [a barebone reproducible example](https://github.com/PedroReyes/wallet_connect_demo) following [the details we can find here](https://docs.walletconnect.com/web3modal/nextjs/about#trigger-the-modal). It is important to remark that the errors appears when I set my project ID.

![image](https://github.com/WalletConnect/walletconnect-monorepo/assets/4096108/e48dccc1-e91a-4bb7-87ca-b42aa2339e4f)

**Reply by PedroReyes** [reply]
Hello @glitch-txs ,

let me know if you want my project ID to have exactly the same issue I have in local. You can either contact me on discord or I pass it here and I will use a new one on production.

**Reply by glitch-txs** [reply]
hi sorry for the delay, it seems that in your example the comments break the bundler, after removing them everything works as expected with no errors:

![image](https://github.com/WalletConnect/walletconnect-monorepo/assets/66949816/6c92dfe2-0dd7-4be6-ab70-1f4bbfc4bc7b)

![image](https://github.com/WalletConnect/walletconnect-monorepo/assets/66949816/b3e765dd-9d36-47fe-aaa1-16824d00a55d)

**Reply by glitch-txs** [reply]
could you try by disabling your chrome extensions?

**Reply by PedroReyes** [reply]
Hello @glitch-txs ,

I have updated the last commit removing the comments.

I have disabled all my extension except metamask.

The errors persists.

![image](https://github.com/WalletConnect/walletconnect-monorepo/assets/4096108/fa7515a7-02ce-405a-bf0a-7ca3fa850dee)

What else could be the difference between our environments?

Output with all extensiosn disabled (and a reset) in Google Chrome:

![image](https://github.com/WalletConnect/walletconnect-monorepo/assets/4096108/884c75a7-8c47-46a8-a8c8-af2c4bebf435)


Output with all extensions disabled (and a reset) in Microsoft Edge:

![image](https://github.com/WalletConnect/walletconnect-monorepo/assets/4096108/e75a1220-3308-4786-b70c-bfefc4b962d8)

I have no clue why this is happening 😣

**Reply by glitch-txs** [reply]
really weird, works for me with no errors on edge too. Could you try creating a new project ID?

**Reply by PedroReyes** [reply]
Okey dokey, I'll be back to my PC in a few hours. I will give it a try then. That's probably the variable causing all this. I'll let you know.

Thanks @glitch-txs for giving me support on weekends. Have a nice Sunday 🤗

**Reply by PedroReyes** [reply]
Hello @glitch-txs ,

we are moving forward. The issue seems to be the project ID. However, the problem has not been solved.

I have created two new project IDs (one with the URL project verified and one without verification. 

In both cases the WalletConnect successfully connects in localhost and in both cases WalletConnect does not even pop-up in production. 

However, in localhost, with the verified project I receive the error we have been talking about the last days and with the project ID non-verified I don't get any error in localhost.

Might be that somehow this error is coming from some sort of bug or mismanagement of my project URL in WalletConnect systems?

**Reply by PedroReyes** [reply]
Hello @glitch-txs ,

What do you think about the use case presented? Might be an issue with my URL in WalletConnect systems?

Is there any other explanation?

**Reply by glitch-txs** [reply]
About the modal not poping up, that's not related to the project ID but there might be something in your code or bundler that's affecting the modal.
For the other error, if you have your domain verified the error shouldn't appear on production.

So far I haven't been able to repro the issue

**Reply by PedroReyes** [reply]
Hello @glitch-txs ,

I have pushed in my last commit [here](https://github.com/PedroReyes/wallet_connect_demo) the project ID from WalletConnect. You will be able to reproduce the issue and see that is related to the project ID:

![image](https://github.com/WalletConnect/walletconnect-monorepo/assets/4096108/f2bdf85e-320e-4e91-bfaf-6c61493c1791)

Now definitely you should be able to reproduce the issue. Let me know if you have any problem.

Best regards,
Pedro Reyes.

**Reply by glitch-txs** [reply]
I can see the errors, the first one is fine and shouldn't appear on production. The second one I can only see it on Edge but it's not causing any problems to connect or to open the modal. 
Uploaded it to Vercel and works well, it's getting CORS from Cloud but the modal is opeing correctly for me: https://wallet-connect-demo-teal.vercel.app/

**Reply by PedroReyes** [reply, marked as answer]
Hello @glitch-txs ,

We can close the ticket. I didn't figure out yet the core problem but I could solve it by disabling one of the functionalities (traditional gmail login). I will look more in detail this part of the code that I have disabled and see what was interacting with WalletConnect.

Thanks for your support ✌️

Best regards,
Pedro Reyes.
